File: exam-system-java/server/src/main/java/com/yyjy/exam/question/api/QuestionController.java
package com.yyjy.exam.question.api;

import cn.dev33.satoken.annotation.SaCheckRole;
import com.yyjy.exam.common.convention.result.R;
import com.yyjy.exam.entity.question.dto.*;
import com.yyjy.exam.entity.question.entity.Questions;
import com.yyjy.exam.entity.question.io.req.*;
import com.yyjy.exam.entity.question.io.res.QuestionGenerateDto;
import com.yyjy.exam.entity.question.io.res.QuestionPageRes;
import com.yyjy.exam.question.service.QuestionService;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.babyfish.jimmer.client.meta.Api;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

@SaCheckRole("admin")
@Api
@RestController
@RequiredArgsConstructor
@RequestMapping("/question")
public class QuestionController {
	
	private final QuestionService questionService;
	
	/**
	 * 新增题目
	 */
	@Api
	@PostMapping("/add")
	public R<Questions> addQuestion(@RequestBody @Valid QuestionSaveRequest question) {
		return R.ok(questionService.save(question));
	}
	
	/**
	 * 更新题目
	 */
	@Api
	@PutMapping("/update")
	public R<Questions> updateQuestion(@RequestBody QuestionUpdateInput question) {
		return R.ok(questionService.update(question));
	}
	
	/**
	 * 删除题目
	 */
	@Api
	@DeleteMapping("/remove/{id}")
	public R<Void> removeQuestion(@PathVariable long id) {
		questionService.remove(id);
		return R.ok();
	}
	
	/**
	 * 获取题目详情
	 */
	@Api
	@GetMapping("/{id}")
	public R<QuestionsPageView> getQuestion(@PathVariable long id) {
		return R.ok(questionService.getById(id));
	}
	
	/**
	 * 获取题目列表
	 */
	@Api
	@GetMapping("/list")
	public R<QuestionPageRes> listQuestions(@Validated QuestionListReq req) {
		
		return R.ok(questionService.list(req));
	}
	
	/**
	 * 获取热门题目 获取访问次数最多的热门题目，用于首页推荐展示
	 */
	@Api
	@GetMapping("/popular")
	public R<List<QuestionsPageView>> getPopularQuestions(
			@RequestParam(defaultValue = "10") int size
	) {
		return R.ok(questionService.getPopularQuestions(size));
	}
	
	/**
	 * 热门题目缓存初始化、数据重置
	 */
	@Api
	@PostMapping("/popular/refresh")
	public R<Void> refreshQuestions() {
		questionService.refreshPopularQuestions();
		return R.ok();
	}
	
	/**
	 * 下载Excel导入模板
	 */
	@Api
	@GetMapping("/batch/template")
	public void downloadTemplate(HttpServletResponse response) {
		questionService.exportTemplate(response);
	}
	
	/**
	 * 预览Excel文件内容
	 */
	@Api
	@PostMapping("/batch/preview-excel")
	public R<List<QuestionImportView>> previewExcel(@RequestPart("file") MultipartFile file) {
		return R.ok(questionService.parseExcel(file));
	}
	
	/**
	 * AI智能生成题目
	 */
	@Api
	@PostMapping("/batch/ai-generate")
	public R<List<QuestionGenerateDto>> aiGenerate(@RequestBody QuestionGenerateReq req) {
		return R.ok(questionService.generateQuestions(req));
	}
	
	/**
	 * 批量导入题目
	 */
	@Api
	@PostMapping("/batch/import-questions")
	public R<String> importQuestions(@RequestBody List<QuestionImportInput> questions) {
		return R.ok(questionService.importBatch(questions));
	}
	
	/**
	 * Excel批量导入题目
	 */
	@Api
	@PostMapping("/batch/excel-import-questions")
	public R<String> importExcelQuestions(@RequestPart("file") MultipartFile file) {
		return R.ok(questionService.importBatchExcel(file));
	}
}

File: exam-system-server/src/main/kotlin/com/yyjy/controller/QuestionController.kt
package com.yyjy.controller

import com.yyjy.common.R
import com.yyjy.models.dto.QuestionGenerateDto
import com.yyjy.models.entity.Questions
import com.yyjy.models.entity.dto.*
import com.yyjy.models.req.QuestionGenerateReq
import com.yyjy.models.req.QuestionListReq
import com.yyjy.models.res.QuestionPageRes
import com.yyjy.repository.QuestionsRepository
import com.yyjy.service.QuestionService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.servlet.http.HttpServletResponse
import org.babyfish.jimmer.client.meta.Api
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile

@Api
@Tag(name = "题目模块")
@RequestMapping("/question")
@RestController
class QuestionController(
    private val questionService: QuestionService,
    private val questionRepository: QuestionsRepository
) {

    @Api
    @Operation(summary = "新增题目", description = "添加新的考试题目，支持选择题、判断题、简答题等多种题型")
    @PostMapping("/add")
    fun addQuestion(@RequestBody question: QuestionSaveInput): R<Questions> {

        return R.ok(questionService.save(question))
    }

    @Api
    @Operation(summary = "更新题目")
    @PutMapping("/update")
    fun updateQuestion(@RequestBody question: QuestionUpdateInput): R<Questions> {
        return R.ok(questionService.update(question))
    }

    @Api
    @Operation(summary = "删除题目")
    @DeleteMapping("/remove/{id}")
    fun removeQuestion(@PathVariable id: Long): R<String?> {
        questionService.remove(id)
        return R.ok()
    }

    @Api
    @Operation(summary = "获取题目", description = "获取指定ID的题目完整信息，包括题目内容、选项、答案等详细数据")
    @GetMapping("/{id}")
    fun getQuestion(@PathVariable id: Long): R<QuestionsPageView?> {
        val question = questionService.getById(id)
        return R.ok(question)
    }

    @Api
    @Operation(
        summary = "题目列表",
        description = "根据传入的多个筛选条件，返回分页后的题目列表。结果中需包含题目的选项和答案等详细信息。"
    )
    @GetMapping("/list")
    fun listQuestions(@Validated req: QuestionListReq): R<QuestionPageRes> {

        val questions: QuestionPageRes = questionService.list(req)

        return R.ok(questions)
    }

    @Api
    @Operation(summary = "获取热门题目", description = "获取访问次数最多的热门题目，用于首页推荐展示")
    @GetMapping("/popular")
    fun getPopularQuestions(size: Int = 10): R<List<QuestionsPageView>> {
        return R.ok(questionService.getPopularQuestions(size))
    }

    @Api
    @Operation(summary = "热门题目缓存初始化、数据重置")
    @PostMapping("/popular/refresh")
    fun refreshQuestions(): R<List<QuestionsPageView>> {
        questionService.refreshPopularQuestions()
        return R.ok()
    }

    @Api
    @Operation(summary = "下载Excel导入模板", description = "获取标准模板文件")
    @GetMapping("/batch/template")
    fun downloadTemplate(response: HttpServletResponse) {
        questionService.exportTemplate(response)
    }

    @Api
    @Operation(summary = "预览Excel文件内容", description = "解析但不入库")
    @PostMapping("/batch/preview-excel")
    fun previewExcel(// 强烈推荐使用 @RequestPart 而不是 @RequestParam
        @RequestPart("file")
        @Parameter(description = "要上传的文件", required = true)
        file: MultipartFile
    ): R<List<QuestionImportView>> {
        return R.ok(questionService.parseExcel(file))
    }

    @Api
    @Operation(summary = "AI智能生成题目", description = "根据指定主题和要求智能生成题目，支持预览后再决定是否导入")
    @PostMapping("/batch/ai-generate")
    fun aiGenerate(@RequestBody req: QuestionGenerateReq): R<List<QuestionGenerateDto>?> {

        return R.ok(questionService.generateQuestions(req))
    }

    @Api
    @Operation(
        summary = "批量导入题目",
        description = "将题目列表批量导入到数据库，支持Excel解析后的导入或AI生成后的确认导入"
    )
    @PostMapping("/batch/import-questions")
    fun importQuestions(@RequestBody question: List<QuestionImportInput>): R<String?> {
        return R.ok(questionService.importBatch(question))
    }

    @Api
    @Operation(summary = "Excel批量导入题目", description = "解析Excel文件并将题目批量导入到数据库")
    @PostMapping("/batch/excel-import-questions")
    fun importExcelQuestions(
        @RequestPart("file")
        @Parameter(description = "要上传的文件", required = true)
        file: MultipartFile
    ): R<String?> {
        return R.ok(questionService.importBatchExcel(file))
    }
}
